### PR TITLE
kubeadm: improve error message about node subnet size validation

### PR DESCRIPTION
**Pull request by pacoxu**
**What type of PR is this?**
/kind cleanup
**What this PR does / why we need it**:

> [root@master-dce4 ~]# kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr="172.30.0.0/26,fdff:ffff:ffff:ffff:0:0::/96"   --service-cidr="172.31.0.0/16,fdff:ffff:ffff:ffff:0:1::/108" --dry-run
> [podSubnet: **Invalid value: "172.30.0.0/26": pod subnet size is smaller than the node subnet size**, podSubnet: **Invalid value: "fdff:ffff:ffff:ffff::/96": pod subnet size is smaller than the node subnet size**]

When I try to install kubeadm with dual-stack enabled, I got the error above. The subnet is smaller than node subnet size. But it doesn't print the node subnet size.

**Which issue(s) this PR fixes**:
part of https://github.com/kubernetes/kubeadm/issues/1612
- will working on kubeadm dual-stack 1.21 supports.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
None
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 1]
```diff
@@ -440,7 +440,7 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
 		// the pod subnet mask needs to allow one or multiple node-masks
 		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
 		if maskSize > nodeMask {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(),  fmt.Sprintf("pod subnet size is smaller than the node subnet size %d", nodeMask)))
```
Comment (aojea) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:443]: what do you think about being more pedantic?
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(),  fmt.Sprintf("pod subnet size %s is smaller than the node subnet size %d", maskSize, nodeMask)))
```

same in the error below

**Comment by pacoxu** [comment]
squashed the commit per your comments @aojea

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 2]
```diff
@@ -440,7 +440,7 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
 		// the pod subnet mask needs to allow one or multiple node-masks
 		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
 		if maskSize > nodeMask {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(),  fmt.Sprintf("pod subnet size %s is smaller than the node subnet size %d", maskSize, nodeMask)))
 		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
 			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask and node-mask difference can not be greater than %d", constants.PodSubnetNodeMaskMaxDiff)))
```
Comment (aojea) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:445]: since we are doing this, we can fix this one too

Comment (pacoxu) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Walk through the file and find another one. Fixed them all now.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 3]
```diff
@@ -440,9 +440,9 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
 		// the pod subnet mask needs to allow one or multiple node-masks
 		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
 		if maskSize > nodeMask {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet size %d is smaller than the node subnet size %d", maskSize, nodeMask)))
```
Comment (SataQiu) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:443]: Emm.. `mask size` and `subnet size` are different. The bigger the mask, the smaller the subnet.
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf(" the size of pod subnet with mask %d is smaller than the size of node subnet with mask %d", maskSize, nodeMask)))
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 4]
```diff
@@ -440,9 +440,9 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
...
 		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask and node-mask difference can not be greater than %d", constants.PodSubnetNodeMaskMaxDiff)))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask %d and node-mask %d difference can not be greater than %d", nodeMask, maskSize, constants.PodSubnetNodeMaskMaxDiff)))
```
Comment (SataQiu) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:445]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask %d and node-mask %d difference can not be greater than %d", maskSize, nodeMask, constants.PodSubnetNodeMaskMaxDiff)))
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 5]
```diff
@@ -397,7 +397,7 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
 	for _, s := range subnets {
 		numAddresses := utilnet.RangeSize(s)
 		if numAddresses < minAddrs {
-			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), "subnet is too small"))
+			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), fmt.Sprintf("subnet %d is too small, at least %d", numAddresses, minAddrs)))
```
Comment (SataQiu) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:400]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), fmt.Sprintf("subnet with %d address(es) is too small, at least %d", numAddresses, minAddrs)))
```

**Review by SataQiu — COMMENTED** [review]
Thanks @pacoxu!
add some comments.

**Comment by SataQiu** [comment]
FYI: Similar PR #98785

**Comment by pacoxu** [comment]
> #98785

@SataQiu  Thanks for your comments.
It makes sense to me. I accepted your comments and squashed the commits.

@driazati Sorry for dup fix on the same thing. Feel free to include/combine this PR in your PR if you like, and close this after #98785 is merged. There are some review comments here that might make sense.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 6]
```diff
@@ -440,9 +440,9 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
 		// the pod subnet mask needs to allow one or multiple node-masks
 		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
 		if maskSize > nodeMask {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf(" the size of pod subnet with mask %d is smaller than the size of node subnet with mask %d", maskSize, nodeMask)))
```
Comment (SataQiu) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:443]: Oh...  There seems to be an extra space before `the size of ...`.

Comment (pacoxu) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: 🤦‍♂️

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 7]
```diff
@@ -397,7 +397,7 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
 	for _, s := range subnets {
 		numAddresses := utilnet.RangeSize(s)
 		if numAddresses < minAddrs {
-			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), "subnet is too small"))
+			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), fmt.Sprintf("subnet with %d address(es) is too small, at least %d", numAddresses, minAddrs)))
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:400]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, s.String(), fmt.Sprintf("subnet with %d address(es) is too small, the minimum is %d", numAddresses, minAddrs)))
```

is this better?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 8]
```diff
@@ -440,9 +440,9 @@ func ValidatePodSubnetNodeMask(subnetStr string, c *kubeadm.ClusterConfiguration
 		// the pod subnet mask needs to allow one or multiple node-masks
 		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
 		if maskSize > nodeMask {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("the size of pod subnet with mask %d is smaller than the size of node subnet with mask %d", maskSize, nodeMask)))
 		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
-			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask and node-mask difference can not be greater than %d", constants.PodSubnetNodeMaskMaxDiff)))
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask %d and node-mask %d difference can not be greater than %d", maskSize, nodeMask, constants.PodSubnetNodeMaskMaxDiff)))
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:445]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask (%d) and node-mask (%d) difference is greater than %d", maskSize, nodeMask, constants.PodSubnetNodeMaskMaxDiff)))
```

i'd reword it like so given the values are provided here.

**Review by neolit123 — COMMENTED** [review]
thanks
/lgtm
/approve

**Comment by pacoxu** [comment]
Thanks for all your comments. 
- I think I need an English training.  😭

**Comment by aojea** [comment]
> * I think I need an English training.  sob

hahah I think that none of the reviewers is english native, so let´s merge before one sees this and correct us all ;)

**Comment by aojea** [comment]
?? what happen @pacoxu ??

**Comment by pacoxu** [comment]
squashed the commit

**Comment by neolit123** [comment]
> hahah I think that none of the reviewers is english native, so let´s merge before one sees this and correct us all ;)

yeah, all of us are doing a best effort on that front :+1:

**Comment by neolit123** [comment]
/retest
looks like a pkg/kubelet flake.

**Review by SataQiu — COMMENTED** [review]
/lgtm
